code/_helpers/icons.dm

```diff
@@ -791,10 +768,17 @@ The _flatIcons list is a cache for generated icon files.
 		else
 			iconmode = blendMode2iconMode(curblend)
 		// Blend the overlay into the flattened icon
-		flat.Blend(add, iconmode, I:pixel_x + 2 - flatX1, I:pixel_y + 2 - flatY1)
+		flat.Blend(add, iconmode, I.pixel_x + 2 - flatX1, I.pixel_y + 2 - flatY1)
 
 	if(A.color)
-		flat.Blend(A.color, ICON_MULTIPLY)
+		if(islist(A.color)) // Probably a matrix.
+			// This is not actually the correct arg order, but I don't know how to translate
+			// colour matrix to MapColors(). It's close enough for government work considering
+			// that the only thing in the game that uses a matrix is the Hawaiian shirt.
+			flat.MapColors(arglist(A.color))
+		else if(istext(A.color)) // Probably a valid color.
+			flat.Blend(A.color, ICON_MULTIPLY)
+
 	if(A.alpha < 255)
```

Comment:
So if you want to be technically correct, you should move this line to only apply if color is a string with 7 characters. In all other cases, the alpha has already been captured in `color` and applied in the code above.